testRequirement.coffee

```diff
@@ -62,35 +70,67 @@ matchesSubstring = (objectKey, substringTest, object) ->
   substringValues = substringTest['substring'] # array of test substrings
   substringValues = if util.isArray substringValues then substringValues else [ substringValues ] #put substringValues in array if it isn't already
 
-  objectValue = object[objectKey] # array of document field values
+  objectValue = object[objectKey] # document field values
   objectValues = if util.isArray objectValue then objectValue else [ objectValue ] #put objectValues in array if it isn't already
 
   for objectValue in objectValues
+    objectValueStr = JSON.stringify(objectValue) # look for substrings in stringified arrays and objects too
+    match = false
     for substring in substringValues
       if 'ignoreCase' not of substringTest or substringTest['ignoreCase'] # ignore case of string by default and when ignoreCase is true
         re = RegExp substring, "im" # case insensitive, multiline matching
       else
         re = RegExp substring, "m" # multiline matching
-      if objectValue
-        objectValue = objectValue.match re
+      if objectValueStr
+        match = objectValueStr.match re
+
+      # if substring in object value str
+      return true if match
+  return false
+
+# Check if the test is an empty test
+isEmptyTest = (testValues) ->
+  if (_.isPlainObject testValues) and 'empty' of testValues then true else false
+
+# Check if specific field value is empty
+isEmpty = (objectKey, emptyTest, object) ->
+  objectValue = object[objectKey]
+  shouldBeEmpty = emptyTest['empty']
+
+  if shouldBeEmpty #return true if objectValue is empty
+    return objectValue.length is 0
+
+  else if not shouldBeEmpty #return true if objectValue is not empty
+    return objectValue.length isnt 0
 
-      # if substring in object value
-      return true if objectValue
+# Check if the test is a pubdate test
+isPubdateTest = (testValues) ->
+  if (_.isPlainObject testValues) and 'year' of testValues then true else false
+
+# Extract specific time from the document pubdate field and compare it with the test
+matchesPubdate = (objectKey, pubdateTest, object) ->
+  objectValue = object[objectKey]
+  year = new Date(objectValue * 1000).getFullYear() # convert to milliseconds
```

Comment:
This is brittle, doesn't let `objectValue` be in ms or s. The Date class is actually already smart enough to detect this. You can replace `objectValue * 1000` with just `objectValue`